lighthouse-core/audits/seo/canonical.js

```diff
@@ -0,0 +1,166 @@
+ */
+'use strict';
+
+const Audit = require('../audit');
+const LinkHeader = require('http-link-header');
+const URL = require('../../lib/url-shim');
+const LINK_HEADER = 'link';
+
+/**
+ * @param {string} headerValue
+ * @returns {Array<string>}
+ */
+function getCanonicalLinksFromHeader(headerValue) {
+  const linkHeader = LinkHeader.parse(headerValue);
+
+  return linkHeader.get('rel', 'canonical').map(c => c.uri);
+}
+
+/**
+ * @param {string} headerValue
+ * @returns {Array<string>}
+ */
+function getHreflangsFromHeader(headerValue) {
+  const linkHeader = LinkHeader.parse(headerValue);
+
+  return linkHeader.get('rel', 'alternate').map(h => h.uri);
+}
+
+/**
+ * Returns true if given string is a valid absolute or relative URL
+ * @param {string} url
+ * @returns {boolean}
+ */
+function isValidRelativeOrAbsoluteURL(url) {
+  try {
+    new URL(url, 'https://example.com/');
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+/**
+ * Returns a primary domain for provided URL (e.g. http://www.example.com -> example.com).
+ * Note that it does not take second-level domains into account (.co.uk).
+ * @param {URL} url
+ * @returns {string}
+ */
+function getPrimaryDomain(url) {
+  return url.hostname.split('.').slice(-2).join('.');
+}
+
+class Canonical extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      name: 'canonical',
+      description: 'Document has a valid `rel=canonical`',
+      failureDescription: 'Document does not have a valid `rel=canonical`',
+      helpText: 'Canonical links suggest which URL to show in search results. ' +
+        'Read more in [Use canonical URLs]' +
+        '(https://support.google.com/webmasters/answer/139066).',
+      requiredArtifacts: ['Canonical', 'Hreflang'],
+    };
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+
+    return artifacts.requestMainResource(devtoolsLogs)
+      .then(mainResource => {
+        const baseURL = new URL(mainResource.url);
+        let canonicals = [];
+        let hreflangs = [];
+
+        mainResource.responseHeaders
+          .filter(h => h.name.toLowerCase() === LINK_HEADER)
+          .forEach(h => {
+            canonicals = canonicals.concat(getCanonicalLinksFromHeader(h.value));
+            hreflangs = hreflangs.concat(getHreflangsFromHeader(h.value));
+          });
+
+        canonicals = canonicals.concat(artifacts.Canonical);
+
+        artifacts.Hreflang.forEach(({href}) => hreflangs.push(href));
+
+        hreflangs = hreflangs
+          .filter(href => isValidRelativeOrAbsoluteURL(href))
+          .map(href => (new URL(href, baseURL)).href); // normalize URLs
+
+        if (canonicals.length === 0) {
+          return {
+            rawValue: true,
```

Comment:
when 4052 lands, we'll want to add `notApplicable: true` to this case.

Reply:
I added it right away (it doesn't really matter if this PR will be merge after or before #4052)